### PR TITLE
Implement `process_attester_slashings`

### PR DESCRIPTION
**Pull request by hwwhww**
### What was wrong?
- Part of #238
- Implement `process_attester_slashings` and related validations
- Spec reference: https://github.com/ethereum/eth2.0-specs/blob/26908d55790fcb24bdf89c13957a54bba3d75e22/specs/core/0_beacon-chain.md#attester-slashings-1

### How was it fixed?
1. Implement `process_attester_slashings`
2. Implement `validate_attester_slashing` for validating each `AttesterSlashing`
3. Add `validate_attester_slashing` tests in `tests/eth2/beacon/state_machines/forks/test_serenity_block_attester_slashing_validation.py`

4. Address https://github.com/ethereum/eth2.0-specs/issues/651

#### TODO:
- [x] Add test for `process_attester_slashing`

#### Cute Animal Picture

nature-animal-wildlife-wild-zoo-fur-red-brown-mammal-fauna-kangaroo-close-up-australia-furry-vertebrate-marsupial-outback-macropodidae-macropus-macropod-horse-like-mammal-587557

### CODE REVIEW

File: eth2/beacon/state_machines/forks/serenity/operation_processing.py [hunk 1]
```diff
@@ -42,6 +43,46 @@ def process_proposer_slashings(state: BeaconState,
...
+
+        current_epoch = state.current_epoch(config.EPOCH_LENGTH)
+        slashable_indices = tuple(
+            index
+            for index in attester_slashing.slashable_attestation_1.validator_indices
+            if index in (
+                attester_slashing.slashable_attestation_2.validator_indices and
+                state.validator_registry[index].slashed_epoch > current_epoch
+            )
+        )
```
Comment (djrtwo) [inline eth2/beacon/state_machines/forks/serenity/operation_processing.py]: Need to validate that `len(slashable_indices) > 0` here.
"Verify that len(slashable_indices) >= 1" in the spec

File: tests/eth2/beacon/state_machines/forks/test_serenity_operation_processing.py [hunk 2]
```diff
@@ -126,6 +128,69 @@ def test_process_proposer_slashings(genesis_state,
...
+        new_state = process_attester_slashings(
+            state,
+            block,
+            config,
+        )
+        # Check if slashed
+        assert (
+            new_state.validator_balances[attester_index] < state.validator_balances[attester_index]
+        )
+    # else:
```
Comment (djrtwo) [inline tests/eth2/beacon/state_machines/forks/test_serenity_operation_processing.py]: Should we fill some failure cases in before merge?

File: eth2/beacon/state_machines/forks/serenity/block_validation.py [hunk 3]
```diff
@@ -571,6 +648,11 @@ def verify_slashable_attestation_signature(state: 'BeaconState',
         SignatureDomain.DOMAIN_ATTESTATION,
     )
 
+    # No custody bit 1 indice
+    if len(all_indices[1]) == 0:
+        pubkeys = pubkeys[:1]
+        messages = messages[:1]
+
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply eth2/beacon/state_machines/forks/serenity/block_validation.py]: Yes, my intention is reducing some calculations here. (At lease reduce one `hash_to_G2`)

[the remark this answers is not in the document]
Comment (ralexstokes) [reply eth2/beacon/state_machines/forks/serenity/block_validation.py]: gotcha -- works for me

File: tests/eth2/beacon/state_machines/forks/test_serenity_operation_processing.py (hunk 2 above)
Comment (hwwhww) [reply tests/eth2/beacon/state_machines/forks/test_serenity_operation_processing.py]: Oh thanks! That was some leftover. Added a light test!

**Pull request merged**